src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/reports/ReconciliationService.java

```diff
@@ -0,0 +1,142 @@
+        this.receivedZipFileConverter = receivedZipFileConverter;
+    }
+
+    public List<Discrepancy> getReconciliationReport(ReconciliationStatement reconciliationStatement) {
+        Map<Pair<String, String>, ReportedZipFile> reportedZipFilesMap =
+            reconciliationStatement.envelopes
+                .stream()
+                .collect(
+                    toMap(
+                        envelope -> Pair.of(envelope.zipFileName, envelope.container), identity()
+                    )
+                );
+
+        List<ReceivedZipFile> receivedZipFiles =
+            receivedZipFileRepository.getReceivedZipFilesReportFor(reconciliationStatement.date);
+        List<ReceivedZipFileData> receivedZipFileDataList =
+            receivedZipFileConverter.convertReceivedZipFiles(receivedZipFiles);
+
+        List<Discrepancy> discrepancies = new ArrayList<>();
+
+        receivedZipFileDataList
+            .forEach(receivedZipFile -> {
+                Pair<String, String> receivedZipFileKey =
+                    Pair.of(receivedZipFile.zipFileName, receivedZipFile.container);
+                if (reportedZipFilesMap.containsKey(receivedZipFileKey)) {
+                    ReportedZipFile reportedZipFile = reportedZipFilesMap.get(receivedZipFileKey);
+                    compareLists(
+                        discrepancies,
+                        receivedZipFile,
+                        reportedZipFile.paymentDcns,
+                        receivedZipFile.paymentDcns,
+                        PAYMENT_DCNS_MISMATCH
+                    );
+                    compareLists(
+                        discrepancies,
+                        receivedZipFile,
+                        reportedZipFile.scannableItemDcns,
+                        receivedZipFile.scannableItemDcns,
+                        SCANNABLE_DOCUMENT_DCNS_MISMATCH
+                    );
+                } else {
+                    discrepancies.add(
+                        new Discrepancy(
+                            receivedZipFile.zipFileName,
+                            receivedZipFile.container,
+                            RECEIVED_BUT_NOT_REPORTED.text
+                        )
+                    );
+                }
+            });
+
+        findReportedButNotReceived(reportedZipFilesMap, receivedZipFileDataList, discrepancies);
+
+        return discrepancies;
+    }
+
+    private void findReportedButNotReceived(
+        Map<Pair<String, String>, ReportedZipFile> reportedZipFilesMap,
+        List<ReceivedZipFileData> receivedZipFileDataList,
+        List<Discrepancy> discrepancies
+    ) {
+        Set<Pair<String, String>> receivedZipFilesSet =
+            receivedZipFileDataList
+                .stream()
+                .map(receivedZipFile -> Pair.of(receivedZipFile.zipFileName, receivedZipFile.container))
+                .collect(toSet());
+        reportedZipFilesMap.keySet()
```

Comment:
`ReportedZipFile` doesn't play anything here. perhaps we can pass key set straight away?